UX FormCollection (#90)
symfony/ux: src/FormCollection/README.md

# UX Form Collection

Symfony UX Form collection is a Symfony bundle providing light UX for collection
in Symfony Forms.

## Installation

UX Form Collection requires PHP 7.2+ and Symfony 4.4+.

Install this bundle using Composer and Symfony Flex:

```sh
composer require symfony/ux-form-collection

# Don't forget to install the JavaScript dependencies as well and compile
yarn install --force
yarn encore dev
```

Also make sure you have at least version 2.0 of [@symfony/stimulus-bridge](https://github.com/symfony/stimulus-bridge)
in your `package.json` file.

You need to select the right theme from the one you are using :
```yaml
# config/packages/twig.yaml
twig:
  # For bootstrap for example
  form_themes: ['@FormCollection/form_theme_div.html.twig']
```
You have 2 different themes :
- `@FormCollection/form_theme_div.html.twig`

> - `@FormCollection/form_theme_table.html.twig`

Reviewer: This is the biggest question make for me with this very nice PR. If we provide a built-in form theme to "rendering everything for them", then we will also need a Bootstrap 4/5 theme... and a tailwind theme. And, if the user needs to customize how things look, then they need to override a fairly complex form theme that we've created here.

I'm not *definitely* against this. However, I'd like to see if we can document how this new feature could be used if there *were* no form themes included. For example,

> Now that your `BlogFormType` form set up and with a `comments` field that is a `CollectionType`, you can
> render it in your template:

```twig
{% macro commentFormRow(commentForm) %}
    <div
        class="col-4"
        data-symfony--ux-form-collection--collection-target="entry"
    >
        {{ form_errors(commentForm) }}
        {{ form_row(commentForm.content) }}
        {{ form_row(commentForm.otherField) }}

        <a data-action="symfony--ux-form-collection--collection#delete">
            Remove
        </a>
    </div>
{% endblock %}

<div
    class="row"
    {{ stimulus_controller('symfony/ux-form-collection/collection', {
        prototype: _self.commentFormRow(form.comments.vars.prototype),
    }) }}
>
    {% for commentForm in form.comments %}
        {{ _self.commentFormRow(commentForm) }}
    {% endfor %}

    <a data-action="symfony--ux-form-collection--collection#add">
        Add Another
    </a>
</div>
```

This would give me full control over everything, including the buttons. The Stimulus controller would just handle making it all work :). The worst pieces are the super-ugly `data-action` and `data-...-target` attributes. But this would look much nicer if we merged https://github.com/symfony/webpack-encore-bundle/pull/124

There may be some other missing pieces... but in theory, this should be enough to work. The "delete" button should not need a `data-index-entry` attribute, as we can look to its ancestors for the "closest" entry target to know which item is being removed.

So, this is my thinking :). No form theme, but we make it as dead-simple as possible for users to render things themselves.

Author: My idea was to make a component very easy to use.
It's true that template overloading can be complicated due to attributes related to Stimulus.
With these 2 templates (`@FormCollection/form_theme_div.html.twig`, `@FormCollection/form_theme_table.html.twig`) we cover a maximum of use cases.

For your proposal, there should be explicit errors when the controller is wrong or the action is not correctly entered.

I think you need to know Stimulus to be able to set up HTML according to the documentation.

Reviewer: To make a better judgement, I'll need some time to play with this in a real project :). Would you be willing/able to push a demo app to GitHub that uses this (with some quick setup instructions)?

Author: Here you have my repo for my test :
https://github.com/stakovicz/ux-collection-test/
Enjoy !

Author: I've made an update.
Now you can use a predefined theme or not.
You have the choice.

I need help to do more tests.

Reviewer 2: Hi @weaverryan,

Any reason to use a macro instead of a proper form theme here?

The macro forces you to use it everywhere it's needed (eg. prototype & entries as in your example) while a "real" form theme would apply automatically to both.

[Check the Symfony doc](https://symfony.com/doc/4.4/form/form_themes.html) for the different ways to set themes in Symfony.

## Usage

The most common usage of Form Collection is to use it as a replacement of
the native CollectionType class:

```php
// ...
use Symfony\UX\FormCollection\Form\CollectionType;

class BlogFormType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            // ...

>             ->add('comments', CollectionType::class, [

Reviewer: Maybe the class should be called `UXCollectionType` - that will avoid auto-complete confusion with the original.

                // ...
                'button_add' => [
                     // Default text for the add button
                    'text' => 'Add',    
                    // Default attr class for the add button
                    'attr' => ['class' => 'btn btn-outline-primary'] 
                ],
                'button_delete' => [
                    // Default text for the delete button
                    'text' => 'Remove',    
                    // Default class for the delete button
                    'attr' => ['class' => 'btn btn-outline-secondary']    
                ],
            ])
            // ...
        ;
    }

    // ...
}

> ```

Reviewer: You should also show how to render it (I know, it's just "rendering the field", but people will still wonder)

```
You have 2 different themes :
- `@FormCollection/form_theme_div.html.twig`
- `@FormCollection/form_theme_table.html.twig`

[Check the Symfony doc](https://symfony.com/doc/4.4/form/form_themes.html) for the different ways to set themes in Symfony.

## Usage

The most common usage of Form Collection is to use it as a replacement of
the native CollectionType class:

```php
// ...
use Symfony\UX\FormCollection\Form\UXCollectionType;

class BlogFormType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            // ...
            ->add('comments', UXCollectionType::class, [
                // ...
                'button_add' => [
                     // Default text for the add button
                    'text' => 'Add',    
                    // Add HTML classes to the add button
                    'class' => 'btn btn-outline-primary'
                ],
                'button_delete' => [
                    // Default text for the delete button
                    'text' => 'Remove',    
                    // Add HTML classes to the add button
                    'class' => 'btn btn-outline-secondary'

>                 ],

Reviewer 3: I'm not sure about me but have you tried with 2 ButtonType in the Form when the options "allow_add" & "allow_delete" are true?
This would allow to modify their appearance via the twig blocks

>                 ],

Author: @maxhelias allow_add is a boolean type from the original [FormType](https://symfony.com/doc/current/reference/forms/types/collection.html#allow-add).

> # UX Form Collection

Reviewer: We're now doing this as an `index.rst` file. So, it just needs to move, then convert to RST

Author: I don't see any RST files in the other parts 🤨. They are all in .md

Reviewer 2: Your branch may be a little out-of-date, see eg. https://github.com/symfony/ux/tree/2.x/src/Chartjs/Resources/doc

Author: Thanks, I had not looked in this file 😬

> Symfony UX Form collection is a Symfony bundle providing light UX for collection

Reviewer: ```suggestion
Symfony UX Form collection is a Symfony bundle providing JavaScript-powered "add" and "remove"
buttons to your `CollectionType` fields.
```

It'd be GREAT to have a little screenshot :)

in Symfony Forms.

## Installation

UX Form Collection requires PHP 7.2+ and Symfony 4.4+.

Install this bundle using Composer and Symfony Flex:

```sh
composer require symfony/ux-form-collection

# Don't forget to install the JavaScript dependencies as well and compile
yarn install --force
yarn encore dev
```

Also make sure you have at least version 2.0 of [@symfony/stimulus-bridge](https://github.com/symfony/stimulus-bridge)
in your `package.json` file.

## Use predefined theme

> You need to select the right theme from the one you are using:

Reviewer: ```suggestion
First, choose a form theme to use for styling your collection type:
```

```yaml
# config/packages/twig.yaml
twig:
    # For bootstrap for example

>     form_themes: ['@FormCollection/form_theme_div.html.twig']

Reviewer: I think we should only have one form theme (`form_theme_div.html.twig`). If we DID have a 2nd form theme, I would make it a bootstrap theme so that the user could instantly get something nice-looking inside of Bootstrap... but I'm not sure that's necessary because it's really easy to add `btn btn-primary` classes to the buttons from within the form.

Anyways, if the user needs to heavily customize their buttons, let's let them create their own form theme to override our's. Also, I think we should prevent the user from needing to do this step at all. We can automatically add this form theme for them: https://github.com/symfony/ux/blob/09c9effcc2582d4de0907c9e12c35afa313433ae/src/Dropzone/DependencyInjection/DropzoneExtension.php#L27-L37 (don't forget the `PrependExtensionInterface`.

```

There are 2 predefined themes available:

-   `@FormCollection/form_theme_div.html.twig`
-   `@FormCollection/form_theme_table.html.twig`

[Check the Symfony doc](https://symfony.com/doc/4.4/form/form_themes.html) for the different ways to set themes in Symfony.

> ## Use a custom form theme

Reviewer: Let's move this section down, below `Usage`. And we can call it: `Customizing how the Field Renders`.

Consider your `BlogFormType` form set up and with a comments field that is a `CollectionType`, you can
render it in your template:

```twig
{% macro commentFormRow(commentForm) %}
    <div
        class="col-4"

>         data-symfony--ux-form-collection--collection-target="entry"

Reviewer: We should show `{{ stimulus_controller() }}` here, and `{{ stimulus_action() }}` and `{{ stimulus_target() }} in the other places.

    >
        {{ form_errors(commentForm) }}
        {{ form_row(commentForm.content) }}
        {{ form_row(commentForm.otherField) }}

        <button type="button" data-action="symfony--ux-form-collection--collection#delete">
            Remove
        </button>
    </div>
{% endmacro %}

<div
    class="row"
    data-controller="symfony--ux-form-collection--collection"
    data-prototype="{{ _self.commentFormRow(form.comments.vars.prototype)|e }}"
>
    {% for commentForm in form.comments %}
        {{ _self.commentFormRow(commentForm) }}
    {% endfor %}

    <button type="button" data-action="symfony--ux-form-collection--collection#add">
        Add Another
    </button>
</div>
```

## Usage

The most common usage of Form Collection is to use it as a replacement of
the native CollectionType class:

```php
// ...
use Symfony\UX\FormCollection\Form\UXCollectionType;

class BlogFormType extends AbstractType
{
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            // ...

>             ->add('comments', UXCollectionType::class, [

Reviewer: To be as helpful as possible, I think we need to show the user a simple `CommentType` class. And then, use that here with `'entry_type' => `CommentType::class`.

Overall, this is how my example looks locally - I think these options are helpful to all show!

```php
'allow_add' => true,
'allow_delete' => true,
'entry_type' => FileType::class,
// ideally UXCollectionType can set this for us
'by_reference' => false,
'entry_options' => [
    // hides the "0", "1", "2" labels on the embedded forms
    'label' => false,
],
'button_add_options' => [
    'label' => '<span class="fa fa-plus"></span>',
    'label_html' => true,
    'class' => 'btn btn-secondary btn-sm'
],
'button_delete_options' => [
    'label' => '<span class="fa fa-times"></span>',
    'label_html' => true,
    'class' => 'btn btn-danger btn-sm'
],
```

                // ...
                // Default text for the add button (used by predefined theme)
                'button_add_text' => 'Add',
                // Add HTML classes to the add button (used by predefined theme)

>                 'button_add_class' => 'btn btn-outline-primary',

Reviewer 4: ~~Would more go with `button_add_attr` I think no other form types yet have option to configure `class` explicit instead they always had an option to render `_attr`.~~

As shown beneath I would go with `button_add_options` which contains all button options also the `label` and so on which is also rendered as a prototype.

Author: I made the changes based on your feedback.
Next, the modifications on the buttons.

                'button_delete_options' => [
                    'label' => 'Remove',                    // Default text for the delete button (used by predefined theme)
                    'class' => 'btn btn-outline-secondary', // Add HTML classes to the add button (used by predefined theme)
                ],
            ])
            // ...
        ;
    }

    // ...
}
```

You can display it using Twig as you would normally with any form:

```twig
{# edit.html.twig #}

{{ form_start(form) }}
    {# ... #}
    {{ form_row(comments) }}
    {# ... #}
{{ form_end(form) }}

> ```

Reviewer: Somewhere, we need to mention that, in order for this to work, they likely need `orphanRemoval` and `cascade=['persist']` on their entities. This is what my field looks like:

```php
#[ORM\OneToMany(mappedBy: 'directory', targetEntity: File::class, orphanRemoval: true, cascade: ['persist'])]
private $files;
```

Symfony UX Form collection is a Symfony bundle providing light UX for collection
in Symfony Forms.

## Installation

UX Form Collection requires PHP 7.2+ and Symfony 4.4+.

Install this bundle using Composer and Symfony Flex:

```sh
composer require symfony/ux-form-collection

# Don't forget to install the JavaScript dependencies as well and compile
yarn install --force
yarn encore dev
```

Also make sure you have at least version 2.0 of [@symfony/stimulus-bridge](https://github.com/symfony/stimulus-bridge)
in your `package.json` file.

## Use predefined theme

You need to select the right theme from the one you are using :
```yaml
# config/packages/twig.yaml
twig:
  # For bootstrap for example
  form_themes: ['@FormCollection/form_theme_div.html.twig']
```
You have 2 different themes :
- `@FormCollection/form_theme_div.html.twig`
- `@FormCollection/form_theme_table.html.twig`

[Check the Symfony doc](https://symfony.com/doc/4.4/form/form_themes.html) for the different ways to set themes in Symfony.

## Use manual theming

> Consider your `BlogFormType` form set up and with a comments field that is a `CollectionType`, you can
render it in your template:

```twig

> {% macro commentFormRow(commentForm) %}

Author: It's @weaverryan's idea. It is easier and more flexible to implement.

    <div
        class="col-4"
        data-symfony--ux-form-collection--collection-target="entry"
    >
        {{ form_errors(commentForm) }}
        {{ form_row(commentForm.content) }}
        {{ form_row(commentForm.otherField) }}

        <button type="button" data-action="symfony--ux-form-collection--collection#delete">

>             Remove

Reviewer 2: ```suggestion
            {{ button_delete_text|trans }}
```

        </button>
    </div>
{% endmacro %}

<div
    class="row"
    data-controller="symfony--ux-form-collection--collection"
    data-prototype="{{ _self.commentFormRow(form.comments.vars.prototype)|e }}"
>
    {% for commentForm in form.comments %}
        {{ _self.commentFormRow(commentForm) }}
    {% endfor %}

    <button type="button" data-action="symfony--ux-form-collection--collection#add">

>         Add Another

Reviewer 2: ```suggestion
        {{ button_add_text|trans }}
```
